File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第2章付账问题.py
#付账问题 2018年第九届蓝桥杯C/C++大学A组，Java大学A组
#题目链接： lanqiaoOJ 174
from math import *
n, s = map(int,input().split())
a = list(map(int,input().split()))
a.sort()
avg = s/n
sum = 0
for i in range(n):
     if a[i]*(n-i) < s:
          sum += pow(a[i]-avg,2)
          s -= a[i]
     else:
          cur_avg = s/(n-i);      #更新平均出钱数
          sum += pow(cur_avg-avg,2)*(n-i)
          break
print("{:.4f}".format(sqrt(sum/(n))))

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第2章付账问题.cpp
//付账问题 2018年第九届蓝桥杯C/C++大学A组，Java大学A组
//题目链接： lanqiaoOJ 174 

#include <bits/stdc++.h>
using namespace std;
const int M = 5e5;
long long a[M];
int main(){
    int n;  long long s;
    scanf("%d %ld",&n,&s);
    for(int i=1;i<=n;i++)  scanf("%ld",&a[i]);
    sort(a+1,a+n+1);       //排序，从小到大
    double avg = 1.0*s/n;  //平均值
    double sum = 0.0;
    for(int i=1;i<=n;i++){
        if(a[i]*(n+1-i) < s){
//需要把钱全拿出的人：（1）钱不够平均数的，（2）钱够平均数，但也不是很多的
            sum += (a[i]-avg)*(a[i]-avg);
            s -= a[i];            //更新剩余钱数
        }
        else{       //不用把钱全拿出的人：非常有钱，不管怎么平均都够
            double cur_avg = 1.0*s/(n+1-i);    //更新平均出钱数
            sum += (cur_avg-avg)*(cur_avg-avg)*(n+1-i);
            break;
        }
    }
    printf("%.4f",sqrt(sum/n));
    return 0;
}
